developer-extension/src/panel/hooks/useEvents/facetRegistry.ts

```diff
@@ -0,0 +1,94 @@
+import { getEventSource, type SdkEvent } from '../../sdkEvent'
+import { createLogger } from '../../../common/logger'
+import { FACETS, type FacetId, type FacetValue } from '../../facets.constants'
+
+const logger = createLogger('facetRegistry')
+
+type FieldPath = string
+type FieldValue = string | number | null | boolean
+type FieldMultiValue = FieldValue | FieldValue[]
+
+export class FacetRegistry {
+  facetValueCounts: Map<FacetId, Map<FacetValue, number>> = new Map(FACETS.map((facet) => [facet.id, new Map()]))
+  eventFacetsCache: WeakMap<SdkEvent, Map<FacetId, FacetValue>> = new WeakMap()
+
+  addEvent(event: SdkEvent) {
+    const fields = getAllFields(event)
+    const eventSource = getEventSource(event)
+    const facetsCache = new Map<FacetId, FacetValue>()
+
+    for (const facet of FACETS) {
+      let facetValue: FacetValue
+      if (facet.id === '$eventSource') {
+        facetValue = eventSource
+      } else {
+        const [, facetEventSource, fieldPath] = /^\$(\w+)\.(.*)$/.exec(facet.id)!
+        if (facetEventSource !== eventSource) {
+          continue
+        }
+        const fieldValue = fields.get(fieldPath)
+        if (typeof fieldValue !== 'string') {
+          continue
+        }
+        facetValue = fieldValue
+      }
+
+      const valueCounts = this.facetValueCounts.get(facet.id)!
+      valueCounts.set(facetValue, (valueCounts.get(facetValue) ?? 0) + 1)
+
+      facetsCache.set(facet.id, facetValue)
+    }
+
+    this.eventFacetsCache.set(event, facetsCache)
+  }
+
+  getFacetValueForEvent(event: SdkEvent, facetId: FacetId): FacetValue | undefined {
+    return this.eventFacetsCache.get(event)?.get(facetId)
+  }
+
+  getFacetValueCounts(facetId: FacetId): Map<FacetValue, number> {
+    return this.facetValueCounts.get(facetId)!
+  }
+
+  clear() {
+    this.facetValueCounts = new Map(FACETS.map((facet) => [facet.id, new Map()]))
+  }
+}
+
+export function getAllFields(event: object) {
+  const fields = new Map<FieldPath, FieldMultiValue>()
+
+  getAllFieldsRec(event, undefined)
+
+  return fields
+
+  function getAllFieldsRec(value: unknown, path: string | undefined) {
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** The logic is a bit hard to follow. It might be worth splitting or commenting it‏

Reply:
I added examples so we don't need to read the tests, split it a bit, and added some doc. Let me know if it's clearer.